cpp/src/strings/copying/concatenate.cu

```diff
@@ -41,50 +41,66 @@ std::unique_ptr<column> concatenate( std::vector<strings_column_view> const& str
+    thrust::host_vector<column_device_view> h_device_views;
+    for( auto&& scv : strings_columns )
+    {
+        device_cols.emplace_back(column_device_view::create(scv.parent(),stream));
+        h_device_views.push_back(*(device_cols.back()));
+    }
+    rmm::device_vector<column_device_view> device_views(h_device_views);
+    auto execpol = rmm::exec_policy(stream);
+    auto d_views = device_views.data().get();
+    // compute size of the output chars column
+    size_t total_bytes = thrust::transform_reduce( execpol->on(stream),
+        d_views, d_views + device_views.size(),
+        [] __device__ (auto d_view) {
+            if( d_view.size()==0 )
+                return static_cast<size_t>(0);
+            auto d_offsets = d_view.child(strings_column_view::offsets_column_index).template data<int32_t>();
+            size_t size = d_offsets[d_view.size()+d_view.offset()] - d_offsets[d_view.offset()];
+            return size;
+        }, static_cast<size_t>(0), thrust::plus<size_t>());
     CUDF_EXPECTS( total_bytes < std::numeric_limits<size_type>::max(), "total size of strings is too large for cudf column" );
 
     // create chars column
     auto chars_column = make_numeric_column( data_type{INT8}, total_bytes, mask_state::UNALLOCATED, stream, mr);
-    auto chars_view = chars_column->mutable_view();
-    auto d_new_chars = chars_view.data<char>();
+    auto d_new_chars = chars_column->mutable_view().data<char>();
 
     // create offsets column
     auto offsets_column = make_numeric_column( data_type{INT32}, strings_count + 1, mask_state::UNALLOCATED, stream, mr);
     auto offsets_view = offsets_column->mutable_view();
     auto d_new_offsets = offsets_view.data<int32_t>();
 
     // copy over the data for all the columns
-    ++d_new_offsets;
-    size_type offset_adjust = 0;
-    size_type null_count = 0;
+    ++d_new_offsets; // skip the first element which will be set to 0 after the for-loop
+    int32_t offset_adjust = 0; // each section of offsets must be adjusted
+    size_type null_count = 0;  // add up the null counts
     for( auto column = strings_columns.begin(); column != strings_columns.end(); ++column )
     {
-        size_type size = column->size();
-        if( size==0 )
-            continue;
+        size_type column_size = column->size();
+        if( column_size==0 ) // nothing to do
+            continue; // empty column may not have children
+        size_type column_offset = column->offset();
         column_view offsets_child = column->offsets();
         column_view chars_child = column->chars();
-
         // copy the offsets column
         auto d_offsets = offsets_child.data<int32_t>();
-        CUDA_TRY(cudaMemcpyAsync( d_new_offsets, d_offsets+1+column->offset(), size*sizeof(int32_t), cudaMemcpyDeviceToDevice, stream ));
-
+        CUDA_TRY(cudaMemcpyAsync( d_new_offsets, d_offsets+1+column_offset, column_size*sizeof(int32_t), cudaMemcpyDeviceToDevice, stream ));
```

Comment:
The offsets child column is of type `INT32` so the correct native type is `int32_t`

Reply:
But if we change `size_type` to 64-bit as is being discussed this will break.

Reply:
Actually if we change `size_type` to 64-bit then this will not break since the offsets column is created using INT32 data_type. If `size_type` was used here and then suddenly changed to 64-bit even though it was created as an INT32 data_type column then that would be bad.

Reply:
You are correct... I guess maybe we need an index type for column data that corresponds to size_type.

Reply:
There is no 'size_type' column `data_type`. If we want the offsets column to use 64-bit values once size_type has been converted, then we would just need to change the strings' offsets column to use INT64 and int64_t appropriately.